File: core/SymmetricTree/SymmetricTree.py
from typing import Optional
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def isSymmetric(root: Optional[TreeNode]) -> bool:
    queue = deque([root])
    while queue:
        length = len(queue)
        left, right = 0, length - 1
        while left < right:
            if queue[left] and queue[right] and queue[left].val != queue[right].val:
                return False
            if (not queue[left] and queue[right]) or (queue[left] and not queue[right]):
                return False
            left += 1
            right -= 1
        for _ in range(length):
            node = queue.popleft()
            if node:
                queue.append(node.left)
                queue.append(node.right)

    return True

File: core/SymmetricTree/SymmetricTree.java
package com.dbc;

import java.util.*;

public class SymmetricTree {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public boolean isSymmetric(TreeNode root) {
        List<TreeNode> queue = new ArrayList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            int length = queue.size();
            int left = 0, right = length - 1;
            while (left < right) {
                if ((queue.get(left) == null && queue.get(right) != null) || (queue.get(left) != null && queue.get(right) == null)) return false;
                if (queue.get(left) != null && queue.get(right) != null && queue.get(left).val != queue.get(right).val) return false;
                left++;
                right--;
            }
            for (int i = 0; i < length; i++) {
                TreeNode node = queue.remove(0);
                if (node != null) {
                    queue.add(node.left);
                    queue.add(node.right);
                }
            }
        }
        return true;
    }
}
